Signal/src/Models/SyncPushTokensJob.swift

```diff
@@ -2,68 +2,72 @@
 
         let runPromise: Promise<Void> = DispatchQueue.main.promise {
-            // Required to potentially prompt user for notifications settings
-            // before `requestPushTokens` will return.
-            self.pushManager.validateUserNotificationSettings()
+            // HACK: no-op dispatch to work around a bug in PromiseKit/Swift which won't compile
+            // when dispatching complex Promise types. We should eventually be able to delete the 
+            // following two lines, skipping this no-op dispatch.
+            return
         }.then {
-            self.requestPushTokens()
+            return self.pushRegistrationManager.requestPushTokens()
         }.then { (pushToken: String, voipToken: String) in
+            Logger.info("\(self.TAG) finished: requesting push tokens")
             var shouldUploadTokens = false
 
             if self.preferences.getPushToken() != pushToken || self.preferences.getVoipToken() != voipToken {
                 Logger.debug("\(self.TAG) Push tokens changed.")
                 shouldUploadTokens = true
             } else if !self.uploadOnlyIfStale {
-                Logger.debug("\(self.TAG) Uploading even though tokens didn't change.")
+                Logger.debug("\(self.TAG) Forced uploading, even though tokens didn't change.")
                 shouldUploadTokens = true
             }
 
-            Logger.warn("\(self.TAG) lastAppVersion: \(AppVersion.instance().lastAppVersion), currentAppVersion: \(AppVersion.instance().currentAppVersion)")
             if AppVersion.instance().lastAppVersion != AppVersion.instance().currentAppVersion {
-                Logger.debug("\(self.TAG) Fresh install or app upgrade.")
+                Logger.info("\(self.TAG) Uploading due to fresh install or app upgrade.")
                 shouldUploadTokens = true
             }
 
             guard shouldUploadTokens else {
-                Logger.warn("\(self.TAG) Skipping push token upload. pushToken: \(pushToken), voipToken: \(voipToken)")
+                Logger.info("\(self.TAG) No reason to upload pushToken: \(pushToken), voipToken: \(voipToken)")
                 return Promise(value: ())
             }
 
-            Logger.warn("\(self.TAG) Sending new tokens to account servers. pushToken: \(pushToken), voipToken: \(voipToken)")
-
+            Logger.warn("\(self.TAG) uploading tokens to account servers. pushToken: \(pushToken), voipToken: \(voipToken)")
             return self.accountManager.updatePushTokens(pushToken:pushToken, voipToken:voipToken).then {
-                return self.recordNewPushTokens(pushToken:pushToken, voipToken:voipToken)
-            }.then {
-                Logger.debug("\(self.TAG) Successfully ran syncPushTokensJob.")
-            }.catch { error in
-                Logger.error("\(self.TAG) Failed to run syncPushTokensJob with error: \(error).")
+                Logger.info("\(self.TAG) successfully updated push tokens on server")
```

Comment:
This is a promise within a promise?  Couldn't we flatten this?

Reply:
`updatePushTokens` and `recordPushTokensLocally` need to happen async and sequentially, and they both need to have `(pushToken, voipToken)` in scope.

I believe this is the most legible way to achieve that.